smart-contracts/contracts/cl-vault/src/rewards/distribution.rs

```diff
@@ -87,80 +116,119 @@ pub fn handle_collect_spread_rewards_reply(
-        .parse::<u128>()?
-        .into();
-
-    // for each user with locked tokens, we distribute some part of the rewards to them
-    // get all users and their current pre-distribution rewards
-    let user_rewards: Result<Vec<(Addr, CoinList)>, ContractError> = SHARES
-        .range(deps.branch().storage, None, None, Order::Ascending)
-        .map(|v| -> Result<(Addr, CoinList), ContractError> {
-            let (address, user_shares) = v?;
-            // calculate the amount of each asset the user should get in rewards
-            // we need to always round down here, so we never expect more rewards than we have
-            let user_rewards = rewards.mul_ratio(Decimal::from_ratio(user_shares, total_shares));
-            Ok((address, user_rewards))
-        })
-        .collect();
+    // If the Deque is empty, we've finished distributing rewards so we clear the semaphore flag
+    let is_last_distribution = DISTRIBUTION_SNAPSHOT.is_empty(deps.storage)?;
+    if is_last_distribution {
+        IS_DISTRIBUTING.save(deps.storage, &false)?;
+        HAS_FEE_BEEN_CALCULATED.save(deps.storage, &false)?;
+
+        // Subtract all accumulated all distributed rewards from the current rewards
+        CURRENT_REWARDS.update(
+            deps.storage,
+            |mut old_rewards| -> ContractResult<CoinList> {
+                old_rewards.sub(&distributed_rewards)?;
+                Ok(old_rewards)
+            },
+        )?;
+        // Clear distributed rewards state
+        DISTRIBUTED_REWARDS.save(deps.storage, &CoinList::new())?;
+    } else {
+        DISTRIBUTED_REWARDS.save(deps.storage, &distributed_rewards)?;
+    }
 
-    // add or create a new entry for the user to get rewards
-    user_rewards?
-        .into_iter()
-        .try_for_each(|(addr, reward)| -> ContractResult<()> {
-            USER_REWARDS.update(deps.storage, addr, |old| -> ContractResult<CoinList> {
-                if let Some(old_user_rewards) = old {
-                    Ok(reward.add(old_user_rewards)?)
-                } else {
-                    Ok(reward)
-                }
-            })?;
-            Ok(())
-        })?;
-
-    Ok(vec![Attribute::new(
-        "total_rewards_amount",
-        format!("{:?}", rewards.coins()),
-    )])
+    Ok(Response::new()
+        .add_attribute("total_rewards_amount", format!("{:?}", rewards.coins()))
+        .add_attribute("is_last_distribution", is_last_distribution.to_string()))
 }
 
-fn collect_incentives(deps: Deps, env: Env) -> Result<MsgCollectIncentives, ContractError> {
+fn get_collect_incentives_msg(deps: Deps, env: Env) -> Result<MsgCollectIncentives, ContractError> {
     let position = POSITION.load(deps.storage)?;
     Ok(MsgCollectIncentives {
         position_ids: vec![position.position_id],
         sender: env.contract.address.into(),
     })
 }
 
-fn collect_spread_rewards(deps: Deps, env: Env) -> Result<MsgCollectSpreadRewards, ContractError> {
+fn get_collect_spread_rewards_msg(
```

Comment:
unneeded change

Reply:
Just a clearer name definition expressing what is doing and what is getting, rather then performing the collect_spread_rewards operation

Reply:
oh nvm